R-package/R/xgb.DMatrix.R

```diff
@@ -195,10 +300,491 @@ xgb.DMatrix <- function(
+#' @param cache_prefix The path of cache file, caller must initialize all the directories in this path.
+#' @param missing A float value to represents missing values in data.
+#'
+#' Note that, while functions like \link{xgb.DMatrix} can take a generic `NA` and interpret
+#' correctly it for different types like `numeric` and `integer`, if an `NA` value is passed here,
+#' it will not be adapted for different input types.
+#'
+#' For example, in R `integer` types, missing values are represented by integer number `-2147483648`
+#' (since machine 'integer' types do not have an inherent 'NA' value) - hence, if one passes `NA`,
+#' which is interpreted as a floating-point NaN by 'xgb.ExternalDMatrix', these integer missing values
+#' will not be treated as missing. This should not pose any problem for `numeric` types, since they
+#' do have an inheret NaN value.
+#' @param as_quantile_dmatrix Whether to create the resulting object as a QuantileDMatrix.
+#' @return A 'DMatrix' object.\itemize{
+#' \item If passing `as_quantile_dmatrix=TRUE`, will have additional subclass 'xgb.QuantileDMatrix'
+#' (same as the objects returned by calling \link{xgb.QuantileDMatrix}).
+#' \item If passing `as_quantile_dmatrix=FALSE`, will have additional subclass 'xgb.ExternalDMatrix'.
+#' }
+#' @seealso \link{xgb.DataIter}, \link{xgb.ProxyDMatrix}, \link{xgb.QuantileDMatrix}
+#' @examples
+#' library(xgboost)
+#' data(mtcars)
+#'
+#' # this custom environment will be passed to the iterator
+#' # functions at each call. It's up to the user to keep
+#' # track of the iteration number in this environment.
+#' iterator_env <- as.environment(
+#'   list(
+#'     iter = 0,
+#'     x = mtcars[, -1],
+#'     y = mtcars[, 1]
+#'   )
+#' )
+#'
+#' # Data is passed in two batches.
+#' # In this example, batches are obtained by subsetting the 'x' variable.
+#' # This is not advantageous to do, since the data is already loaded in memory
+#' # and can be passed in full in one go, but there can be situations in which
+#' # only a subset of the data will fit in the computer's memory, and it can
+#' # be loaded in batches that are accessed one-at-a-time only.
+#' # The 'proxy_handle' is supplied externally when this function is called,
+#' # and it must be passed to 'xgb.ProxyDMatrix' inside of the call.
+#' iterator_next <- function(iterator_env, proxy_handle) {
+#'   curr_iter <- iterator_env[["iter"]]
+#'   if (curr_iter >= 2) {
+#'     # there are only two batches, so this signals end of the stream
+#'     return(0)
+#'   }
+#'
+#'   if (curr_iter == 0) {
+#'     x_batch <- iterator_env[["x"]][1:16, ]
+#'     y_batch <- iterator_env[["y"]][1:16]
+#'   } else {
+#'     x_batch <- iterator_env[["x"]][17:32, ]
+#'     y_batch <- iterator_env[["y"]][17:32]
+#'   }
+#'
+#'   # Function 'xgb.ProxyDMatrix' must be called manually
+#'   # at each batch with all the appropriate attributes,
+#'   # such as feature names and feature types.
+#'   xgb.ProxyDMatrix(
```

Comment:
Is it possible that we hide this function and pass a closure to the user like the Python code? I think it's easier to use if the user doesn't need to worry about an additional function and an opaque handle. R is a high level language and it would great if we can hide these C artifacts.

Reply:
I don't think a closure is the best option here.

Another option could be to have the user return the result of this `xgb.ProxyDMatrix` (or `NULL` to signal end of batch), but let this be just a simple list and defer processing for later instead of processing things just after `xgb.ProxyDMatrix` is called.

It would however forego the possibility of doing a manual clean up of whatever environment generated the data inside the iterator if the user wishes to do so immediately before the next batch, but perhaps it's a more logical design.

Reply:
> I don't think a closure is the best option here.

Could you please elaborate on this?

Reply:
> > I don't think a closure is the best option here.
> 
> Could you please elaborate on this?

They are akward to use, hard to document, and hard to examine as a user. Functions, environments and lists are more common and easier to use. Or do you think that a closure could bring some advantage here that would not be achieved by e.g. letting the user return the result of `xgb.ProxyDMatrix` ? (assuming that the handle is left out of the signature)

Reply:
I don't have preference on whether a closure or a function should be used, it's just I would like to hide/abstract many things as possible, and sometimes closure is the way to do it. If there are other approaches, I'm happy to take it.

Reply:
Factored out the `proxy_handle` parameter. Now the user needs to call a function `xgb.ProxyDMatrix` inside of the callback (which doesn't take any handle) and return its result.